File: packages/scanner/src/rules.ts
import type { AuditCategory, CategorySummary, Finding, ScanResult, ScanStatus, Severity } from "./types.js";

const SEVERITY_WEIGHTS: Record<Severity, number> = {
  critical: 25,
  high: 15,
  medium: 8,
  low: 3,
  info: 0,
};

const SEVERITY_ORDER: Record<Severity, number> = {
  critical: 1,
  high: 2,
  medium: 3,
  low: 4,
  info: 5,
};

export function sortFindings(findings: Finding[]): Finding[] {
  return [...findings].sort((a, b) => {
    // 1. First by status: fail, warning, pass
    const statusScore = (s: string) => (s === "fail" ? 1 : s === "warning" ? 2 : 3);
    const diffStatus = statusScore(a.status) - statusScore(b.status);
    if (diffStatus !== 0) return diffStatus;

    // 2. Then by severity
    const diffSev = SEVERITY_ORDER[a.severity] - SEVERITY_ORDER[b.severity];
    if (diffSev !== 0) return diffSev;

    return a.title.localeCompare(b.title);
  });
}

export function computeScanScore(findings: Finding[]): {
  score: number;
  status: ScanStatus;
  grade: "A+" | "A" | "B+" | "B" | "C" | "F";
  summary: string;
} {
  let penalty = 0;

  for (const finding of findings) {
    if (finding.status === "fail") {
      penalty += SEVERITY_WEIGHTS[finding.severity];
    } else if (finding.status === "warning") {
      penalty += Math.round(SEVERITY_WEIGHTS[finding.severity] * 0.4);
    }
  }

  // Calculate score bounded between 0 and 100
  const score = Math.max(0, Math.min(100, 100 - penalty));

  let status: ScanStatus;
  let grade: "A+" | "A" | "B+" | "B" | "C" | "F";
  let summary: string;

  if (score >= 90) {
    status = "excellent";
    grade = score >= 95 ? "A+" : "A";
    summary = "Excellente posture de sécurité périmétrique. Vos en-têtes et protocoles respectent rigoureusement les recommandations de l'OWASP.";
  } else if (score >= 75) {
    status = "good";
    grade = score >= 82 ? "B+" : "B";
    summary = "Bonne posture de sécurité globale avec quelques écarts de configuration ou en-têtes défensifs recommandés à consolider.";
  } else if (score >= 50) {
    status = "warning";
    grade = "C";
    summary = "Posture de sécurité intermédiaire avec des vulnérabilités de configuration exposées (CSP ou protections cross-origin absentes).";
  } else {
    status = "critical";
    grade = "F";
    summary = "Risque critique identifié : absence de HTTPS valide ou manque flagrant des principaux mécanismes défensifs contre les injections.";
  }

  return { score, status, grade, summary };
}

export function buildCategorySummaries(
  findings: Finding[]
): Record<AuditCategory, CategorySummary> {
  const categoriesConfig: Record<AuditCategory, { title: string }> = {
    https: { title: "HTTPS & Transport" },
    headers: { title: "En-têtes de sécurité" },
    cookies: { title: "Cookies" },
    mixed_content: { title: "Contenu mixte" },
    forms: { title: "Formulaires" },
    technology: { title: "Technologies exposées" },
  };

  const summaries: Record<string, CategorySummary> = {};

  for (const [catKey, meta] of Object.entries(categoriesConfig)) {
    const cat = catKey as AuditCategory;
    const catFindings = findings.filter((f) => f.category === cat);
    const total = catFindings.length;
    const passed = catFindings.filter((f) => f.status === "pass").length;

    let catScore = 100;
    if (total > 0) {
      let penalty = 0;
      for (const f of catFindings) {
        if (f.status === "fail") {
          penalty += f.severity === "critical" ? 40 : f.severity === "high" ? 25 : 15;
        } else if (f.status === "warning") {
          penalty += 10;
        }
      }
      catScore = Math.max(0, 100 - penalty);
    }

    const worstStatus = catFindings.some((f) => f.status === "fail")
      ? "fail"
      : catFindings.some((f) => f.status === "warning")
      ? "warning"
      : "pass";

    summaries[cat] = {
      category: cat,
      title: meta.title,
      score: catScore,
      status: worstStatus,
      passedCount: passed,
      totalCount: total,
    };
  }

  return summaries as Record<AuditCategory, CategorySummary>;
}

File: packages/python/securio/rules.py
from __future__ import annotations

from typing import Dict, List, Tuple
from .models import CategorySummary, Finding, FindingStatus, ScanStatus, Severity

SEVERITY_WEIGHTS: Dict[Severity, int] = {
    "critical": 25,
    "high": 15,
    "medium": 8,
    "low": 3,
    "info": 0,
}

SEVERITY_ORDER: Dict[Severity, int] = {
    "critical": 1,
    "high": 2,
    "medium": 3,
    "low": 4,
    "info": 5,
}

STATUS_ORDER: Dict[FindingStatus, int] = {
    "fail": 1,
    "warning": 2,
    "pass": 3,
}

CATEGORIES_CONFIG: Dict[str, str] = {
    "https": "HTTPS & Transport",
    "headers": "En-têtes de sécurité",
    "cookies": "Cookies",
    "mixed_content": "Contenu mixte",
    "forms": "Formulaires",
    "technology": "Technologies exposées",
}


def sort_findings(findings: List[Finding]) -> List[Finding]:
    return sorted(
        findings,
        key=lambda f: (
            STATUS_ORDER.get(f.status, 99),
            SEVERITY_ORDER.get(f.severity, 99),
            f.title.lower(),
        ),
    )


def compute_scan_score(findings: List[Finding]) -> Tuple[int, ScanStatus, str, str]:
    penalty = 0
    for finding in findings:
        weight = SEVERITY_WEIGHTS.get(finding.severity, 0)
        if finding.status == "fail":
            penalty += weight
        elif finding.status == "warning":
            penalty += round(weight * 0.4)

    score = max(0, min(100, 100 - penalty))

    if score >= 90:
        status: ScanStatus = "excellent"
        grade = "A+" if score >= 95 else "A"
        summary = (
            "Excellente posture de sécurité périmétrique. Vos en-têtes et "
            "protocoles respectent rigoureusement les recommandations de l'OWASP."
        )
    elif score >= 75:
        status = "good"
        grade = "B+" if score >= 82 else "B"
        summary = (
            "Bonne posture de sécurité globale avec quelques écarts de "
            "configuration ou en-têtes défensifs recommandés à consolider."
        )
    elif score >= 50:
        status = "warning"
        grade = "C"
        summary = (
            "Posture de sécurité intermédiaire avec des vulnérabilités de "
            "configuration exposées (CSP ou protections cross-origin absentes)."
        )
    else:
        status = "critical"
        grade = "F"
        summary = (
            "Risque critique identifié : absence de HTTPS valide ou manque "
            "flagrant des principaux mécanismes défensifs contre les injections."
        )

    return score, status, grade, summary


def build_category_summaries(findings: List[Finding]) -> Dict[str, CategorySummary]:
    summaries: Dict[str, CategorySummary] = {}

    for cat_key, cat_title in CATEGORIES_CONFIG.items():
        cat_findings = [f for f in findings if f.category == cat_key]
        total = len(cat_findings)
        passed = len([f for f in cat_findings if f.status == "pass"])

        cat_score = 100
        if total > 0:
            penalty = 0
            for f in cat_findings:
                if f.status == "fail":
                    penalty += 40 if f.severity == "critical" else (25 if f.severity == "high" else 15)
                elif f.status == "warning":
                    penalty += 10
            cat_score = max(0, 100 - penalty)

        has_fail = any(f.status == "fail" for f in cat_findings)
        has_warning = any(f.status == "warning" for f in cat_findings)
        worst_status: FindingStatus = "fail" if has_fail else ("warning" if has_warning else "pass")

        summaries[cat_key] = CategorySummary(
            category=cat_key,
            title=cat_title,
            score=cat_score,
            status=worst_status,
            passed_count=passed,
            total_count=total,
        )

    return summaries
